### DISCUSSION TITLE
Directly mutate AST in ESLINT.fix() | Custom Eslint Plugin

### DISCUSSION THREAD

**Discussion by bennobuilder** [eslint/eslint#16855] [Q&A] [answered]
Hey there, 
Is it possible to modify the AST in a custom ESLint plugin's `fix()` method 
for a more flexible approach, rather than being restricted 
to the Node of the `RuleListener`? 

For example, to format:
```jsx
const About = () => {
  return (
    <div>
      <p>About</p>
    </div>
  );
};
```
to:

```jsx
const About = () => {
  return (
    <div>
      <p>{about}</p>
    </div>
  );
};

const about = "About";
```
It's a stupid example, but you get the point.
In this example, I would use `JSXAttribute` as `RuleListener` and show a warning.
But if I now want to apply a fix how can I create & inject the `const about = "About";` 
at the different position in the AST.

Thanks ^^

**Comment by bennobuilder** [comment, marked as answer]
ok right I just noticed that I can access the entire AST from the `context`
and then use the `fixer` methods to insert code after a specific node.

**Reply by bennobuilder** [reply]
ok.. its a bit buggy if there are multiple fixes to do like:
```jsx
const About = () => {
  return (
    <div>
      <p>About</p>
      <p>Second About</p>
    </div>
  );
};
```
by inserting the `const xyz = "xyz"` via `insertingTextAfter()` after the last Node of the AST like:
```js
const ast = context.getSourceCode().ast;
const lastNode = ast.body[ast.body.length - 1];
fixer.insertTextAfter(lastNode, "const xyz = ...");
```
But I suppose I'm not supposed to insert elements after some element out of context of the called listener node anyway, right?
So I decided to do an "after run" by first fixing the JSX, storing the to add variables in a key map and in the `Program:exit` listener I insert the variables at the end of the file in a batch.

**Reply by bennobuilder** [reply]
tbh is there a better way or what's the recommendation in such scenario.. thanks ^^

**Reply by mdjermanovic** [reply]
> ok right I just noticed that I can access the entire AST from the `context`

You can also use `parent` property of nodes. It points to the direct parent node, and from there you can access nodes relative to the current node.

> But I suppose I'm not supposed to insert elements after some element out of context of the called listener node anyway, right?

No, fixes don't have to be restricted to the currently traversed node. You can insert code before or after it.

> So I decided to do an "after run" by first fixing the JSX, storing the to add variables in a key map and in the `Program:exit` listener I insert the variables at the end of the file in a batch.

This doesn't seem necessary, fixes are applied after running rules so they don't change the AST right away.

**Reply by bennobuilder** [reply]
ok awesome 👍 

I currently've the issue that if there are multiple nodes to fix in a `.jsx` file that it only fixes the first JSX node but ignores the other ones,
but only if "Extract class names to identifier" is added to fixers (see code) -> I add something after the last node.

My current fixer function
```ts
fix: (fixer) => {
  const fixers: RuleFix[] = [];

  // Fix "Replace class names with identifier"
  fixers.push(fixer.replaceText(node, `className={${identifier}}`));

  // Fix "Extract class names to identifier"
  const ast = context.getSourceCode().ast;
  const lastNode = ast.body[ast.body.length - 1];
  const toInsertCode = `\n\n${buildOutsourcedClassName(
    splitted.classes,
    identifier,
    lastNode.loc.start.column + 1
  )}`;
  fixers.push(fixer.insertTextAfter(lastNode, toInsertCode));

  return fixers;
},
```

I found that when I do "Extract class names to identifier" in an after run (in `Program:exit`) it works.
The only disadvantage is that the `Program:exit` also shows a warning in the IDE.

<details>
<summary>After Run</summary>

```ts
  'Program:exit': (node) => {
    if (Object.keys(extractedTailwindClasses).length > 0) {
      context.report({
        node,
        messageId: 'invalidInline',
        fix: (fixer) => {
          const ast = context.getSourceCode().ast;

          // Add TailwindCss classes to end of the file (in a batch)
          const lastNode = ast.body[ast.body.length - 1];
          const toInsertCode = Object.keys(extractedTailwindClasses).reduce(
            (previousValue, identifier) => {
              const classes = extractedTailwindClasses[identifier];

              // Add new code block with a constant declaration for the extracted Tailwind class
              if (classes != null) {
                previousValue =
                  previousValue +
                  `\n\n${buildOutsourcedClassName(
                    classes,
                    identifier,
                    lastNode.loc.start.column + 1
                  )}`;
              }

              // Remove the extracted Tailwind class entry from the stored list
              delete extractedTailwindClasses[identifier];

              return previousValue;
            },
            ''
          );

          return fixer.insertTextAfter(lastNode, toInsertCode);
        },
      });
    }
  },

```

</details>

Can you spot what I'm missing / doing wrong here. Because I would love to've everything together and not split between different listeners.

<details>
<summary>Failing Test (works with after run)</summary>

```ts

    Expected value to strictly be equal to:
      "
              import React from 'react';···
              const About: React.FC = () => {
                return (
                  <div className={Container}>
                    <p id=\"text1\" className={Text1}>About</p>
                    <p id=\"text2\" className=\"box-content lg:box-border\">Me</p>
                  </div>
                );
              };···
              export default About;·
              const Container = `
                custom
                container
                animate-spin
                first:flex
              `;·
              const Text1 = `
                p-4
                sm:py-5
                sm:px-7
                lg:p-8
              `;
           "
    Received:
      "
              import React from 'react';···
              const About: React.FC = () => {
                return (
                  <div className={Container}>
                    <p id=\"text1\" className=\"sm:py-5 p-4 sm:px-7 lg:p-8 extract-[Text1]\">About</p>        
                    <p id=\"text2\" className=\"lg:box-border box-content\">Me</p>
                  </div>
                );
              };···
              export default About;·
              const Container = `
                first:flex
                animate-spin
                custom
                container
              `;
           "

```

</details>

Note: This is work in progress code

Thanks ^^

**Reply by bennobuilder** [reply]
Probably because they conflict, although the `lastNode` is the latest added `const` statement -> provided AST was updated with latest changes

**Reply by bennobuilder** [reply]
Here is the project though.. in case anyone has an idea, please let me know :)

https://github.com/bennodev19/eslint-plugin-tailwindcss-jsx/tree/develop

**Reply by mdjermanovic** [reply]
> Probably because they conflict, although the `lastNode` is the latest added `const` statement -> provided AST was updated with latest changes

Yes, this is most likely because the fixes are conflicting (inserting text at the same position). Note that AST doesn't change during the traversal. Can you try running eslint on the command line with the [`--fix`](https://eslint.org/docs/latest/use/command-line-interface#--fix) option and see if all the fixes will be applied? `--fix` is multipass. It lints the code, applies non-conflicting fixes, then lints the new code (new AST) and applies new fixes, and so on up to 10 times. `RuleTester` runs only one pass.

**Reply by bennobuilder** [reply]
Thanks for the reply :) 
It actually works in action when ESLint has the chance to perform up to 10 runs. Do you've an idea how I can write tests for this scenario though if the `RuleTester` does only runs once. Or any idea how I can improve the behavior so that only one run is necessary. 
Appreciate your tips ;D

**Reply by mdjermanovic** [reply]
> Do you've an idea how I can write tests for this scenario though if the `RuleTester` does only runs once.

You could write a sequence of tests where `code` is the `output` of the previous test.

For example:

https://github.com/eslint/eslint/blob/43af24a88b939a62880c37d1332b02f677d82f16/tests/lib/rules/prefer-exponentiation-operator.js#L284-L325
